crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -713,9 +707,21 @@ pub async fn diff_proposals<A>(
                     // TODO: here we should really wait until the fragment is sequenced, otherwise
                     //       we would be going ahead and sequencing more fragments that may not be
                     //       needed. For the moment we just linearly back-off.
-                    fragments_num += 1;
-                    if fragments_num > 2 {
-                        tokio::time::sleep(fragments_num * consensus_delay_estimate).await;
+                    tokio::time::sleep(consensus_delay_estimate).await;
+
+                    let success = checkpoint_db
+                        .lock()
+                        .attempt_to_construct_checkpoint(committee);
+
+                    match success {
+                        Err(_) | Ok(false) => {
```

Comment:
I think on of the error cases is that the tx were not executed. In that case we should stop sequencing fragments, and wait for executions.

Reply:
What if we make attempt_to_construct_checkpoint blocking on the execution?

Reply:
I like this as a direction. We should pass in the pending execution process a way to be notified back when all transactions in a call were executed. This will allow us here to be fully reactive, rather than wait and hope.